### PR TITLE
[Serializer] Fix denormalizing a collection of union types

### LINKED ISSUE #56874: [Serializer] Cannot get base type on `int|string` compound type
### Symfony version(s) affected

7.1-RC1

### Description

The Serializer component switched to the new TypeInfo component and now our serialization breaks:

> Symfony\Component\TypeInfo\Exception\LogicException : Cannot get base type on "int|string" compound type.

### How to reproduce

```php
final readonly class AddToCart
{
    /**
     * @param array<string, string|int> $metadata
     */
    public function __construct(
        public string $cartId,
        public array $metadata,
    ) {}
}
```

> Symfony\Component\TypeInfo\Exception\LogicException : Cannot get base type on "int|string" compound type.
>  /Volumes/CS/www/cosmos/vendor/symfony/type-info/Type/UnionType.php:48
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:705
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:830
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Normalizer/AbstractNormalizer.php:371
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:239
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:335
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Serializer.php:238
>  /Volumes/CS/www/cosmos/vendor/symfony/serializer/Serializer.php:143

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #56874
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 1]
```diff
@@ -702,7 +703,11 @@ private function validateAndDenormalize(Type $type, string $currentClass, string
                 }
 
                 if ($collectionValueType) {
-                    $collectionValueBaseType = $collectionValueType instanceof UnionType ? $collectionValueType->asNonNullable()->getBaseType() : $collectionValueType->getBaseType();
+                    try {
+                        $collectionValueBaseType = $collectionValueType instanceof UnionType ? $collectionValueType->asNonNullable()->getBaseType() : $collectionValueType->getBaseType();
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]:
```suggestion
                        $collectionValueBaseType = $collectionValueType instanceof UnionType ? $collectionValueType->getBaseType() : $collectionValueType->getBaseType();
```

Can we shorten it like this?

Comment (HypeMC) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: @xabbuh Not sure what you mean, the left and right side of the expression would then be the same.

Comment (xabbuh) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Ah indeed, so we could even shorten this a bit more, right?

```php
try {
    $collectionValueBaseType = $collectionValueType->getBaseType();
} catch (TypeInfoLogicException) {
    $collectionValueBaseType = Type::mixed();
}
```

Comment (HypeMC) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: If I understand correctly, the point of `asNonNullable()` is to filter out nullable types, eg if you have `int|null`, the base type if `int`. In that case it makes sense.

Comment (xabbuh) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: The `UnionType` class already takes care of that: https://github.com/symfony/symfony/blob/e02306ba638c32d296c06c424ab11c35421c1743/src/Symfony/Component/TypeInfo/Type/UnionType.php#L43

Comment (HypeMC) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Oh, you're right, I totally missed that.

Comment (HypeMC) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Done

**Review by xabbuh — APPROVED** [review]

**Comment by ruudk** [comment]
I see this bug fix did not make it to the 7.2.0 release 😢 Can it be merged so that it can ship in 7.2.1?

**Comment by fabpot** [comment]
Thank you @HypeMC.

**Comment by fabpot** [comment]
@ruudk Sorry, I missed it. I'm pretty sure we will have a 7.1.1 in the coming days.

**Pull request merged**